Tireddy2 patch 7 (#20)
tireddy2/JOSE_HPKE: draft-rha-jose-hpke-encrypt.md

This specification describes two modes of use for HPKE in JWE:

  *  Integrated Encryption mode, where HPKE is used with a single recipient. This setup is conceptually similar to Direct Key Agreement.
  

>   *  Key Encryption mode, where HPKE is used with multiple recipients. This setup is conceptually similar to Key Agreement with Key Wrapping.

Reviewer: ```suggestion
  *  Key Encryption mode, where HPKE may be used with multiple recipients. This setup is conceptually similar to Key Agreement with Key Wrapping.
```

When the alg value or enc value is set to any of algorithms registered by this specification then the 'epk' header parameter MUST be present, and it MUST be a JSON Web Key as defined in {{EK}} of this document.

The "ek" member of an 'epk' will contain the base64url encoded "enc" value produced by the encapsulate operation of the HPKE KEM.

In all serializations, "ct" will be base64url encoded.

If the 'alg' header parameter is set to the 'HPKE-IntEnc' value (as defined in {{IANA}}), HPKE is used in Integrated Encryption mode; otherwise, it is in Key Encryption mode.

Interested readers will observe this is due to all recipients using the same JWE Protected Header when JSON Serializations are used, as described in {{Section 7.2.1 of RFC7516}}.

We provide the following table for additional clarity:

| Name                   | Recipients | Serializations | Content Encryption Key | Similar to
|---
| Integrated Encryption  | 1          | Compact, JSON  | Derived from HPKE      | Direct Key Agreement

> | Key Encryption         | Many       | Compact, JSON  | Encrypted by HPKE      | Key Agreement with Key Wrapping

Reviewer 2: ```suggestion
| Key Encryption         | 1 or More      | Compact, JSON  | Encrypted by HPKE      | Key Agreement with Key Wrapping
```

Reviewer 2: Ilari wrote:

```
| Name                  | Recipients | HPKE Encrypts
|---
| Integrated Encryption | 1          | Message
| Key Encryption        | >=1        | Content Encryption Key
```

^ this does seem simpler to me

> | Key Encryption         | 1          | Compact, JSON  | Encrypted by HPKE      | Key Agreement with Key Wrapping

Reviewer 2: ```suggestion
```

The Open function will, if successful, decrypts "ct".  When decrypted, the result will be either the CEK (when Key Encryption mode is used), or the content (if Integrated Encryption mode is used).  The CEK is the symmetric key used to decrypt the ciphertext.

## Protected Header Parameters

In order to ensure the security properties of HPKE, the following requirements MUST be followed for protected header parameters:

> * For Integrated Encryption, "epk" MUST be present, and MUST contain an Encapsulated JSON Web Key.

Reviewer: This needs to go the according mode section.

Author: Agreed, moved text.

> * For Integrated Encryption, "alg" MUST be "HPKE-IntEnc", as defined in {{IANA}}, and "enc" MUST be an HPKE algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}}.

Author: Done.

> * For Key Encryption, "alg" MUST NOT be present and "enc" MUST be a content encryption algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}}.

Reviewer: And this makes no sense as per my previous comment about making things required to be protected.

As such this section should be dispersed.

Author: Yes, removed section.

### Integrated Encryption

This mode only supports a single recipient.

HPKE is employed to directly encrypt the plaintext, and the resulting ciphertext is included in the JWE ciphertext. 

In Integrated Encryption mode:

*  The "epk" Header Parameter MUST be present, it MUST contain an Encapsulated JSON Web Key and it MUST occur only within the JWE Protected Header.

> *  The "alg" Header Parameter MUST be "HPKE-IntEnc", "enc" MUST be an HPKE algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}} and they MUST occur only within the JWE Protected Header.

Reviewer 2: ```suggestion
*  The "alg" Header Parameter MUST be "dir", "enc" MUST be an HPKE algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}} and they MUST occur only within the JWE Protected Header.
```

*  The JWE Ciphertext MUST be the resulting HPKE ciphertext ('ct' value) encoded using base64url.

*  The JWE Initialization Vector value MUST be absent. 

*  The JWE Authentication Tag MUST be absent.

*  The JWE Encrypted Key MUST be absent.

*  The HPKE "aad" parameter MUST be set to the JWE Additional Authenticated Data encryption parameter defined in Step 14 of Section 5.1 of {{RFC7516}} as input. 

The following example demonstrates the use of Integrated Encryption with Compact Serialization:

~~~
eyJhbGciOiJkaXIiLCJlbmMiOiJIUEtFLUJhc2UtUDI1Ni1TSEEyNTYtQUVTMTI4R0NNIiwiZXBrIjp7Imt0eSI6IkVLIiwiZWsiOiJCR05ranp0MDc2YnNSR2o3OGFYNUF6VF9IRU9KQmJZOXEyWm9fNWU3dGJLMGFQcXU0ZVQxV0kxNmp2UmxacXBNeXFaZlAtUndSNEp3dGF6XzhVOXRoWEEifX0...DG3qygxcMHw3iACy5mX_T4N4EqWc03W0nkTHjMJsC4nb6JS6vVj6wTGdlr5TOSr0ykaoyzpePXEvEyHhvpUwCyQQr6kbGlGuZsrJdUbZ728vmA.
~~~
{: #integrated-encryption-compact align="left" title="Integrated Encryption with Compact Serialization"}

In the above example, the JWE Protected Header value is: 

~~~
{

>   "alg": "HPKE-IntEnc",

Reviewer 2: ```suggestion
  "alg": "dir",
```

  "enc": "HPKE-Base-P256-SHA256-AES128GCM",
  "epk": {
    "kty": "EK",
    "ek": "BGNkjzt076bsRGj78aX5AzT_HEOJBbY9q2Zo_5e7tbK0aPqu4eT1WI16jvRlZqpMyqZfP-RwR4Jwtaz_8U9thXA"
  }
}
~~~

~~~
{
    "protected": "eyJhbGciOiJkaXIiLCJlbmMiOiJIUEtFLUJhc2UtUDI1Ni1TSEEyNTYtQUVTMTI4R0NNIiwiZXBrIjp7Imt0eSI6IkVLIiwiZWsiOiJCTzRFbGZXd0xKRDZWcERza3c5LWxWMm9OMDJ2U1FKTW55ZHk3enhvSVlKZ1kydk9taE44Q1BqSHdRM1NONkhTcnNHNElacVpHVUR3dExKZjBoeHFTWGsifX0",
    "ciphertext": "1ATsw0jshqPrv8CFcm9Rem9Wfi1Ygv30sozlRTtNNzcaaZ828GqP0AXtqQ1Msv8YXI9XZqh81MK3QnlZ7pOBC1BP7j00J1rrHujdb3zvnOpmJg"
}
~~~
{: #integrated-encryption-json align="left" title="Integrated Encryption with JSON Serialization"}

In the above example, the JWE Protected Header value is: 

~~~
{

>   "alg": ""HPKE-IntEnc",

Reviewer 2: ```suggestion
  "alg": "dir",
```

### Integrated Encryption

This mode only supports a single recipient.

HPKE is employed to directly encrypt the plaintext, and the resulting ciphertext is included in the JWE ciphertext. 

In Integrated Encryption mode:

*  The "epk" Header Parameter MUST be present, it MUST contain an Encapsulated JSON Web Key and it MUST occur only within the JWE Protected Header.

*  The "alg" Header Parameter MUST be "HPKE-IntEnc", "enc" MUST be an HPKE algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}} and they MUST occur only within the JWE Protected Header.

> *  The JWE Ciphertext MUST be the base64url encoded 'ct' value.

Reviewer: ```suggestion
*  The JWE Ciphertext MUST be the resulting HPKE ciphertext and tag ('ct' value) encoded using base64url.
```

Author: I propose the following text:
The JWE Ciphertext MUST be the resulting HPKE ciphertext ('ct' value) encoded using base64url. For example, if an AES-GCM scheme is used, the GCM authentication tag is placed at the end of the HPKE ciphertext output.

Reviewer: That's confusing because HPKE produces only ciphertext and tag `ct`, that's one output value, not two. And the internals of HPKE shouldn't be plucked out of HPKE and placed in examples here.

Reviewer 2: We probably want to say something like this once:

`When tag is non empty, it is the base64url encoded bytes of the AEAD used`

Author: I don't get it. JWE Authentication Tag will always be empty with HPKE.

Author: > That's confusing because HPKE produces only "ciphertext and tag `ct`", that's one output value, not two. And the internals of HPKE shouldn't be plucked out of HPKE and placed in examples here.

Yes, It is implied from HPKE that ct includes both ciphertext and authentication tag. I will modify the text as follows:
The JWE Ciphertext MUST be the resulting HPKE ciphertext ('ct' value) encoded using base64url.

Reviewer: Exactly. All my suggestion is trying to do is to assign as source and name the HPKE "ct" output as per HPKE's own language.

> The JWE Ciphertext MUST be the base64url encoded 'ct' value.

> The JWE Ciphertext MUST be the resulting HPKE ciphertext and tag ('ct' value) encoded using base64url.

Author: I removed the "For example,..." line.

> *  The JWE Initialization Vector value MUST be empty. 
> 
> *  The JWE Authentication Tag MUST be empty.
> 
> *  The JWE Encrypted Key MUST be empty.

Reviewer: ```suggestion
*  The JWE Initialization Vector value MUST be absent. 

*  The JWE Authentication Tag MUST be absent.

*  The JWE Encrypted Key MUST be absent.
```

*  The HPKE "aad" parameter MUST be set to the JWE Additional Authenticated Data encryption parameter defined in Step 14 of Section 5.1 of {{RFC7516}} as input. 

The following example demonstrates the use of Integrated Encryption with Compact Serialization:

~~~
eyJhbGciOiJkaXIiLCJlbmMiOiJIUEtFLUJhc2UtUDI1Ni1TSEEyNTYtQUVTMTI4R0NNIiwiZXBrIjp7Imt0eSI6IkVLIiwiZWsiOiJCR05ranp0MDc2YnNSR2o3OGFYNUF6VF9IRU9KQmJZOXEyWm9fNWU3dGJLMGFQcXU0ZVQxV0kxNmp2UmxacXBNeXFaZlAtUndSNEp3dGF6XzhVOXRoWEEifX0...DG3qygxcMHw3iACy5mX_T4N4EqWc03W0nkTHjMJsC4nb6JS6vVj6wTGdlr5TOSr0ykaoyzpePXEvEyHhvpUwCyQQr6kbGlGuZsrJdUbZ728vmA.
~~~
{: #integrated-encryption-compact align="left" title="Integrated Encryption with Compact Serialization"}

In the above example, the JWE Protected Header value is: 

~~~
{

>   "alg": ""HPKE-IntEnc",

Reviewer 2: ```suggestion
  "alg": "dir",
```

Reviewer 2: I don't like this change, it introduces another algorithm that is conceptually very similar to "dir".

Author: It was discussed and agreed in https://github.com/tireddy2/JOSE_HPKE/pull/19.  Please have a look.

Reviewer: It is somewhat calming that "dir" continues to mean an actual shared secret was used and I don't need to deduce from the "enc" value whether "anyone" could've produced this message or just the parties who know the symmetric secret.

    "ciphertext": "1ATsw0jshqPrv8CFcm9Rem9Wfi1Ygv30sozlRTtNNzcaaZ828GqP0AXtqQ1Msv8YXI9XZqh81MK3QnlZ7pOBC1BP7j00J1rrHujdb3zvnOpmJg"
}
~~~
{: #integrated-encryption-json align="left" title="Integrated Encryption with JSON Serialization"}

In the above example, the JWE Protected Header value is: 

~~~
{
  "alg": ""HPKE-IntEnc",
  "enc": "HPKE-Base-P256-SHA256-AES128GCM",
  "epk": {
    "kty": "EK",
    "ek": "BGNkjzt076bsRGj78aX5AzT_HEOJBbY9q2Zo_5e7tbK0aPqu4eT1WI16jvRlZqpMyqZfP-RwR4Jwtaz_8U9thXA"
  }
}
~~~

### Key Encryption

This mode supports more than one recipient.

HPKE is used to encrypt the	Content Encryption Key (CEK), and the resulting ciphertext is included in the JWE ciphertext. 
The plaintext will be encrypted using the CEK as explained in Step 15 of Section 5.1 of {{RFC7516}}.

When there are multiple recipients, the sender MUST place the 'epk' parameter in the per-recipient unprotected header to indicate the use of HPKE. In this case, the 'enc' (Encryption Algorithm) Header Parameter MUST be a content encryption algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}}, and it MUST be present in the JWE Protected Header.

In Key Encryption mode: 

> *  The JWE Encrypted Key MUST be the base64url encoded 'ct' value.
> 
> *  The JWE Initialization Vector MUST be produced as described in { Section 5.1 of RFC7516 }.
> 
> *  The JWE Authentication Tag MUST be produced as described in { Section 5.1 of RFC7516 }.

Reviewer: ```suggestion
*  The JWE Encrypted Key MUST be the resulting HPKE ciphertext and tag ('ct' value) encoded using base64url.
```

These are clear, why do they need to be present?

Author: HPKE uses AEAD and it does not produce ciphertext and tag separately (the tag is appended to the ciphertext).

Reviewer: I'm aware, but this is describing Key Encryption, not Integrated Encryption. In the conceptually similar Key Agreement with Key Wrapping the KW's iv and tag would go into a header and not be the JWE IV or Tag. So this does not need to be called out. 

The fact that they're absent is called in Integrated Encryption correctly but in here there's nothing that needs to be said.

Author: Yes, I removed the text related to JWE IV and Authentication Tag.

    "ciphertext": "1ATsw0jshqPrv8CFcm9Rem9Wfi1Ygv30sozlRTtNNzcaaZ828GqP0AXtqQ1Msv8YXI9XZqh81MK3QnlZ7pOBC1BP7j00J1rrHujdb3zvnOpmJg"
}
~~~
{: #integrated-encryption-json align="left" title="Integrated Encryption with JSON Serialization"}

In the above example, the JWE Protected Header value is: 

~~~
{
  "alg": ""HPKE-IntEnc",
  "enc": "HPKE-Base-P256-SHA256-AES128GCM",
  "epk": {
    "kty": "EK",
    "ek": "BGNkjzt076bsRGj78aX5AzT_HEOJBbY9q2Zo_5e7tbK0aPqu4eT1WI16jvRlZqpMyqZfP-RwR4Jwtaz_8U9thXA"
  }
}
~~~

When Integrated Encryption is represented in General JSON Serialization, `iv`, `tag` and `aad` must not be present, as they are always empty.

### Key Encryption

This mode supports more than one recipient.

HPKE is used to encrypt the	Content Encryption Key (CEK), and the resulting ciphertext is included in the JWE ciphertext. 
The plaintext will be encrypted using the CEK as explained in Step 15 of Section 5.1 of {{RFC7516}}.

The sender MUST place the 'epk' and 'alg' parameters in the per-recipient unprotected header, or shared unprotected header to indicate the use of HPKE. 

> The "enc" (Encryption Algorithm) Header Parameter MUST be present in the JWE Protected Header.	 		

Reviewer: Just because this mode supports more recipients does not mean it always will be used with more than 1 recipient.

Because of this you cannot say that epk and alg MUST be placed in {enter location}.

The only requirements that should be present are for when something MUST be integrity protected; therefore, it MUST occur only within the JWE Protected Header.

That's it, no more needs to be said.

Author: I re-phrased the text to address your comment:
When there are multiple recipients, the sender MUST place the 'epk' and 'alg' parameters in the per-recipient unprotected header or shared unprotected header to indicate the use of HPKE. In this case, the 'enc' (Encryption Algorithm) Header Parameter MUST be present in the JWE Protected Header.

Reviewer: - does "enc" actually have to be protected?
- when there are multiple recipients but they all use the same "alg" value, we can't place that "alg" parameter in a protected header?
- when there is one recipient epk can't be protected?
- I could probably go on...

This entire section is inconsistent and redundant with what it is implied from how the JWE serialization formats, JOSE Header structures, and their processing, already function and are defined.

Author: "enc" has to be protected, it is discussed https://datatracker.ietf.org/doc/html/rfc7516#section-7.2.1 that says "This  primarily means that the "enc" (encryption algorithm) Header Parameter value in the JOSE Header for each recipient and any parameters of that algorithm MUST be the same."

However, I agree that "alg" can also be placed in the protected header and "epk" can be protected in case of one receipt but the proposed text is specific to multiple recipients. I will modify the text as follows: 

When there are multiple recipients, the sender MUST place the 'epk' parameter in the per-recipient unprotected header to indicate the use of HPKE. In this case, the 'enc' (Encryption Algorithm) Header Parameter MUST be a content encryption algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}}, and it MUST be present in the JWE Protected Header.

Reviewer: > "enc" has to be protected, it is discussed [datatracker.ietf.org/doc/html/rfc7516#section-7.2.1](https://datatracker.ietf.org/doc/html/rfc7516#section-7.2.1) that says "This primarily means that the "enc" (encryption algorithm) Header Parameter value in the JOSE Header for each recipient and any parameters of that algorithm MUST be the same."

I don't feel that i agree with one leading to the other. Yes, the "enc" values need to be the same for every recipient, clearly, that's inherent to the general JWE format, but that doesn't mean it has to be protected in a produced JWE. Neither does it mean that we have to re-iterate what's already implicit about JWE formats.

Author: "enc" can be protected or shared unprotected, we want to avoid "shared unprotected".  The integrity protected 'enc' provides protection against an attacker who manipulates the encryption algorithm in the 'enc' parameter. The attack is discussed in https://datatracker.ietf.org/doc/draft-ietf-lamps-cms-cek-hkdf-sha256/.

Reviewer: Than that's worth calling out in the draft.

Author: Yes, added text.

      "encrypted_key": "yDVZLsO7-ecy_GCgEluwn9U723TCHNAzeYRRQPOfpHM",
      "header": {
        "kid": "urn:ietf:params:oauth:jwk-thumbprint:sha-256:adjwW6fyyZ94ZBjGjx_OpDEKHLGfd1ELkug_YmRAjCk",

>         "enc": "HPKE-Base-P256-SHA256-AES128GCM",

Reviewer: ```suggestion
        "alg": "HPKE-Base-P256-SHA256-AES128GCM",
```

Author: Good catch.

}
~~~

~~~
{
  "protected": "eyJlbmMiOiJBMTI4R0NNIiwiZXBrIjp7Imt0eSI6IkVLIiwiZWsiOiJCUFJUS244bVFMNGhOMWF5b2tSOGdrcFR5NUhRbGQ0TjBIWFhCOWNYdGpVSVEzN3pzSkRMN1R1Z1ZrbUQxYUZZVHgtMGJNMHRmeHplakxjdFNES2pNUXMifX0",
  "encrypted_key": "zR0ArfrVVRQ9-X_heDU2riwx36QxLBffRrKAWU-tLC4",
  "iv": "o3v11Hw6gUxUN-pY",
  "ciphertext": "Ny-2IDGHMI3MzVsUAVMGNoKAZfoewTZ1dkAIBikPy4eZUfHW_LPhhKpD6Mf4zYGkhAeLwGgJKjyDoFIj0EuDsEtJ",
  "tag": "0sfzHJvxVoWt02EPxMTh8w"

> }

Reviewer: Likewise, this example is missing an "alg" value.

Author: Good catch, fixed.

~~~

> {: #key-encryption-single-recipient-general-json align="left" title="Key Encryption (single recipient) General JSON Serialization"}

Reviewer: ```suggestion
{: #key-encryption-single-recipient-flattened-json align="left" title="Key Encryption (single recipient) Flattened JSON Serialization"}
```

In the above example, the JWE Protected Header value is: 

~~~
{
  "enc": "A128GCM",
  "epk": {
    "kty": "EK",
    "ek": "BPRTKn8mQL4hN1ayokR8gkpTy5HQld4N0HXXB9cXtjUIQ37zsJDL7TugVkmD1aFYTx-0bM0tfxzejLctSDKjMQs"
  }
}
~~~

~~~
eyJlbmMiOiJBMTI4R0NNIiwiZXBrIjp7Imt0eSI6IkVLIiwiZWsiOiJCSjdyZE5ia3Z3dW5zc2RqdTVXRGtBYXpMYUJYM0lkY0xSankxUkRTQTlzaWowMGpkeWJhSHVRUFR0NlAzMUJpMG5FMmtZV183TF90YUFxQUZLNzVEZWsifX0.xaAa0nFxNJxsQQ5J6EFdzUYROd2aV517o2kZnfwhO7s.AgBYEWTj-EMji17I.Ejwu2iEP4xs3FfGO_zTZYu35glQmUvd_qpHpvB1hNqg6Yz5ek3NsZRGMzd--HYWvABNslxBkRwrkZDXnv_BTgOTj.u0ac86ipoAwUZuYwkaKwNw
~~~

> {: #key-encryption-single-recipient-compact align="left" title="Key Encryption (single recipient) Compact"}

Reviewer: This example is missing an "alg" value.

Author: Good catch, fixed.

The following entries are added to the "JSON Web Signature and Encryption Algorithms" registry:

> - Algorithm Name: HPKE-IntEnc
> - Algorithm Description: Integrated Encryption mode for HPKE.
> - Algorithm Usage Location(s): "alg"
> - JOSE Implementation Requirements: Optional
> - Change Controller: IESG
> - Specification Document(s): [[TBD: This RFC]]
> - Algorithm Analysis Documents(s): TODO
> 

Reviewer 2: ```suggestion
```

The following entries are added to the "JSON Web Signature and Encryption Algorithms" registry:

- Algorithm Name: HPKE-IntEnc
- Algorithm Description: Integrated Encryption mode for HPKE.
- Algorithm Usage Location(s): "alg"
- JOSE Implementation Requirements: Optional
- Change Controller: IESG
- Specification Document(s): [[TBD: This RFC]]
- Algorithm Analysis Documents(s): TODO

- Algorithm Name: HPKE-Base-P256-SHA256-AES128GCM
- Algorithm Description: Cipher suite for JOSE-HPKE in Base Mode that uses the DHKEM(P-256, HKDF-SHA256) KEM, the HKDF-SHA256 KDF and the AES-128-GCM AEAD.

> - Algorithm Usage Location(s): "enc"

Reviewer: ```suggestion
- Algorithm Usage Location(s): "alg, enc"
```

- JOSE Implementation Requirements: Optional
- Change Controller: IESG
- Specification Document(s): [[TBD: This RFC]]
- Algorithm Analysis Documents(s): TODO

- Algorithm Name: HPKE-Base-P384-SHA384-AES256GCM
- Algorithm Description: Cipher suite for JOSE-HPKE in Base Mode that uses the DHKEM(P-384, HKDF-SHA384) KEM, the HKDF-SHA384 KDF, and the AES-256-GCM AEAD.

> - Algorithm Usage Location(s): "enc"

Reviewer: ```suggestion
- Algorithm Usage Location(s): "alg, enc"
```

Author: Fixed

- JOSE Implementation Requirements: Optional
- Change Controller: IESG
- Specification Document(s): [[TBD: This RFC]]
- Algorithm Analysis Documents(s): TODO

- Algorithm Name: HPKE-Base-P521-SHA512-AES256GCM
- Algorithm Description: Cipher suite for JOSE-HPKE in Base Mode that uses the DHKEM(P-521, HKDF-SHA512) KEM, the HKDF-SHA512 KDF, and the AES-256-GCM AEAD.

> - Algorithm Usage Location(s): "enc"

Reviewer: ```suggestion
- Algorithm Usage Location(s): "alg, enc"
```

- JOSE Implementation Requirements: Optional
- Change Controller: IESG
- Specification Document(s): [[TBD: This RFC]]
- Algorithm Analysis Documents(s): TODO

- Algorithm Name: HPKE-Base-X25519-SHA256-AES128GCM
- Algorithm Description: Cipher suite for JOSE-HPKE in Base Mode that uses the DHKEM(X25519, HKDF-SHA256) KEM, the HKDF-SHA256 KDF, and the AES-128-GCM AEAD.

> - Algorithm Usage Location(s): "enc"

Reviewer: ```suggestion
- Algorithm Usage Location(s): "alg, enc"
```

- JOSE Implementation Requirements: Optional
- Change Controller: IESG
- Specification Document(s): [[TBD: This RFC]]
- Algorithm Analysis Documents(s): TODO

- Algorithm Name: HPKE-Base-X25519-SHA256-ChaCha20Poly1305
- Algorithm Description: Cipher suite for JOSE-HPKE in Base Mode that uses the DHKEM(X25519, HKDF-SHA256) KEM, the HKDF-SHA256 KDF, and the ChaCha20Poly1305 AEAD.

> - Algorithm Usage Location(s): "enc"

Reviewer: ```suggestion
- Algorithm Usage Location(s): "alg, enc"
```

Author: Fixed.

- JOSE Implementation Requirements: Optional
- Change Controller: IESG
- Specification Document(s): [[TBD: This RFC]]
- Algorithm Analysis Documents(s): TODO

- Algorithm Name: HPKE-Base-X448-SHA512-AES256GCM
- Algorithm Description: Cipher suite for JOSE-HPKE in Base Mode that uses the DHKEM(X448, HKDF-SHA512) KEM, the HKDF-SHA512 KDF, and the AES-256-GCM AEAD.

> - Algorithm Usage Location(s): "enc"

Reviewer: ```suggestion
- Algorithm Usage Location(s): "alg, enc"
```

Author: Fixed.

- JOSE Implementation Requirements: Optional
- Change Controller: IESG
- Specification Document(s): [[TBD: This RFC]]
- Algorithm Analysis Documents(s): TODO

- Algorithm Name: HPKE-Base-X448-SHA512-ChaCha20Poly1305
- Algorithm Description: Cipher suite for JOSE-HPKE in Base Mode that uses the DHKEM(X448, HKDF-SHA512) KEM, the HKDF-SHA512 KDF, and the ChaCha20Poly1305 AEAD.

> - Algorithm Usage Location(s): "enc"

Reviewer: ```suggestion
- Algorithm Usage Location(s): "alg, enc"
```

Author: Fixed.
